Clean up clock interfaces and remove tickless idle

nashif opened the pull request:
- clock: remove z_ from semi-public APIs
- doc: clocks: link clock APIs from doxygen
- clock: rename z_timer_cycle_get_32 -> clock_cycle_get_32
- kernel: remove tickless idle
- drivers: arcv2_timer0: fix comment

nashif committed:
clock: remove z_ from semi-public APIs

The clock/timer APIs are not application facing APIs, however, similar
to arch_ and a few other APIs they are available to implement drivers
and add support for new hardware and are documented and available to be
used outside of the clock/kernel subsystems.

Remove the leading z_ and provide them as clock_* APIs for someone
writing a new timer driver to use.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: clocks: link clock APIs from doxygen

Use :c:func: where clock APIs are being used and link the group into
RST.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
clock: rename z_timer_cycle_get_32 -> sys_clock_cycle_get_32

This is another API that is being used in all timer drivers and is not
internal to the clock subsystem. Remove the leading z_ and make promote
it to a cross-subsystem API.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: remove tickless idle

This feature predated the tickless kernel and has been in legacy mode
for a while. We now have no drivers or systems that do not support
tickless, so remove this option and cleanup the code to only use
tickless.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
drivers: arcv2_timer0: fix comment

Comment talking about z_timer_int_handler which does not exist, replace
with timer_int_handler.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

dcpleung commented:
Wonder if `sysclk_*` would be better, since this is for system clock (or kernel clock). Might be a good idea to distinguish this from real time clock or wall clock.

pabigot commented on drivers/timer/Kconfig:
```diff
@@ -30,12 +30,12 @@ config APIC_TIMER_IRQ_PRIORITY
 	  TSC frequency to CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC (the local APIC
 	  timer frequency), then enable this for a much faster and more
-	  accurate z_timer_cycle_get_32().
+	  accurate clock_cycle_get_32().
```
I agree with @dcpleung that it would be best to more clearly identify which clock is involved. 

`z_timer_cycle_get_32()` is aliased by `arch_k_cycle_get_32()` which in turn is aliased by `k_cycle_get_32()`.

The rate it advances is provided by `sys_clock_hw_cycles_per_sec()` and some implementations also use `sys_clock_` as a standard prefix in the system clock driver(s) for the architecture.

So it would seem to make sense to have this be `sys_clock_get_32()`, or `sys_clock_cycle_get_32()`.

nashif committed:
clock: z_clock_isr -> sys_clock_isr

Do not use z_ for internal APIs, z_ is for private APIs within one
subsystem only.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
clock: z_clock_device_ctrl -> sys_clock_device_ctrl

Do not use z_ for internal APIs, z_ is for private APIs within one
subsystem only.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sys_lock: extra 'needed' removed

Minor typo fix.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
clock: rename z_clock_hw_cycles_per_sec_runtime_get

Do not use z_ for internal APIs, z_ is for private APIs within one
subsystem only.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
clock: rename z_tick_get_32 ->  sys_clock_tick_get_32

Do not use z_ for internal APIs, z_ is for private APIs within one
subsystem only.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
clocks: rename z_tick_get -> sys_clock_tick_get

Do not use z_ for internal APIs, z_ is for private APIs within one
subsystem only.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
clock: renmae z_timeout_end_calc -> sys_clock_timeout_end_calc

Do not use z_ for internal APIs, z_ is for private APIs within one
subsystem only.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
clock: remove unused z_ms_to_ticks

z_ms_to_ticks is not being used anywhere.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

ruuddw commented on drivers/timer/arcv2_timer0.c:
```diff
@@ -326,7 +326,7 @@ void z_clock_set_timeout(int32_t ticks, bool idle)
 	 * be ignored, as 32-bits timer will overflow in a not-long time.
 	 */
-	if (IS_ENABLED(CONFIG_TICKLESS_IDLE) && ticks == K_TICKS_FOREVER) {
+	if (ticks == K_TICKS_FOREVER) {
```
Shouldn't this become CONFIG_TICKLESS_KERNEL (and the next one as well)?

nashif committed:
hal_ti: update to use k_ms_to_ticks_ceil32

z_ms_to_ticks is deprecated, so use k_ms_to_ticks_ceil32.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

stephanosio commented on drivers/timer/altera_avalon_timer_hal.c:
```diff
@@ -69,3 +74,15 @@ uint32_t sys_clock_cycle_get_32(void)
+	driver_uptime += ticks;
+	sys_clock_announce(ticks);
+}
+#define sys_clock_announce(t) wrapped_announce(t)
```
This `#define` seems to be misplaced.

Since `sys_clock_announce` is not overridable (i.e. not a weak symbol), I would say the best way to handle this would be to explicitly use `wrapped_announce` throughout this file.
